module/User/src/User/Authentication/Storage/Session.php

```diff
@@ -2,25 +2,166 @@
 
 namespace User\Authentication\Storage;
 
-use Zend\Authentication\Storage;
+use Zend\Authentication\Storage,
+    User\Model\Session as SessionModel,
+    Zend\Http\Header\SetCookie;
 
 class Session extends Storage\Session
 {
 
+    /**
+     * @var The service manager
+     */
+    protected $sm;
+
+    /**
+     * @var boolean indicating whether we should remember the user
+     */
+    protected $rememberMe;
+
     public function setRememberMe($rememberMe = 0, $time = 1209600)
     {
-        if ($rememberMe == 1) {
-            $this->session->getManager()->rememberMe($time);
+        $this->rememberMe = $rememberMe;
+        if ($rememberMe) {
+            $this->saveSession($this->session->{$this->member});
         }
     }
 
     public function forgetMe()
     {
-        $this->session->getManager()->forgetMe();
+        $this->rememberMe = false;
+    }
+
+    public function isEmpty()
+    {
+        if (isset($this->session->{$this->member})) {
+            return false;
+        }
+
+        return !$this->readDatabaseSession();
+
+    }
+
+    protected function readDatabaseSession()
+    {
+        $mapper = $this->sm->get('user_mapper_session');
+        $request = $this->sm->get('Request');
+        $cookies = $request->getHeaders()->get('cookie');
+        if (!isset($cookies->SESSID) || !isset($cookies->SECRET)) {
+            return false;
+        }
+        $session = $mapper->find($cookies->SESSID, $cookies->SECRET);
+        if ($session === null) {
+            return false;
+        }
+        $this->session->{$this->member} = $session->getUser()->getLidnr();
+        return true;
+    }
+
+    public function read()
+    {
+        return $this->session->{$this->member};
+    }
+
+    /**
+     * Defined by Zend\Authentication\Storage\StorageInterface
+     *
+     * @param  mixed $contents
+     * @return void
+     */
+    public function write($contents)
+    {
+        $this->session->{$this->member} = $contents;
+        if ($this->rememberMe) {
+            $this->saveSession($contents);
+        }
+    }
+
+    /**
+     * Store the current session.
+     *
+     * @param \User\Model\User $user the logged in user
+     *
+     * @return SessionModel
+     */
+    public function saveSession($user)
+    {
+        $mapper = $this->sm->get('user_mapper_session');
+        $session = new SessionModel();
+        $session->setIp($this->sm->get('user_remoteaddress'));
+        $user = $this->sm->get('user_service_user')->detachUser($user);
+        $session->setUser($user);
+        $session->setSecret($this->generateSecret());
+        $session->setCreatedAt(new \DateTime());
+        $session->setLastActive(new \DateTime());
+        $mapper->persist($session);
+        $mapper->flush();
+        $this->saveCookie($session);
```

Comment:
Maybe include some whitespace here and there for readability? This kinda looks incoherent and just some random lines of code thrown together.